test/testSwCanvas.cpp

```diff
@@ -0,0 +1,86 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <thorvg.h>
+#include "catch.hpp"
+
+using namespace tvg;
+
+
+TEST_CASE("Missing Initialization", "[tvgSwCanvas]")
+{
+    auto canvas = SwCanvas::gen();
+    REQUIRE(canvas == nullptr);
+}
+
+TEST_CASE("Basic Creation", "[tvgSwCanvas]")
+{
+    Initializer::init(CanvasEngine::Sw, 0);
+
+    auto canvas = SwCanvas::gen();
+    REQUIRE(canvas);
+
+    auto canvas2 = SwCanvas::gen();
+    REQUIRE(canvas2);
+
+    auto canvas3 = SwCanvas::gen();
+    REQUIRE(canvas3);
+
+    Initializer::term(CanvasEngine::Sw);
+}
+
+TEST_CASE("Target Buffer", "[tvgSwCanvas]")
+{
+    Initializer::init(CanvasEngine::Sw, 0);
+
+    auto canvas = SwCanvas::gen();
+    REQUIRE(canvas);
+
+    uint32_t buffer[100*100];
+    REQUIRE(canvas->target(buffer, 100, 100, 100, SwCanvas::Colorspace::ABGR8888) == Result::Success);
+    REQUIRE(canvas->target(buffer, 100, 100, 100, SwCanvas::Colorspace::ARGB8888) == Result::Success);
+
+    REQUIRE(canvas->target(nullptr, 100, 100, 100, SwCanvas::Colorspace::ARGB8888) == Result::InvalidArguments);
+    REQUIRE(canvas->target(buffer, 0, 100, 100, SwCanvas::Colorspace::ARGB8888) == Result::InvalidArguments);
+    REQUIRE(canvas->target(buffer, 100, 0, 100, SwCanvas::Colorspace::ARGB8888) == Result::InvalidArguments);
+    REQUIRE(canvas->target(buffer, 100, 100, 0, SwCanvas::Colorspace::ARGB8888) == Result::InvalidArguments);
+    REQUIRE(canvas->target(buffer, 100, 100, 100, (SwCanvas::Colorspace)-1) == Result::InvalidArguments);
+
+    Initializer::term(CanvasEngine::Sw);
+}
+
+TEST_CASE("Memory Pool", "[tvgSwCanvas]")
+{
+    Initializer::init(CanvasEngine::Sw, 0);
+
+    auto canvas = SwCanvas::gen();
+    REQUIRE(canvas);
+
+    REQUIRE(canvas->mempool(SwCanvas::MempoolPolicy::Default) == Result::Success);
+    REQUIRE(canvas->mempool(SwCanvas::MempoolPolicy::Individual) == Result::Success);
+    REQUIRE(canvas->mempool(SwCanvas::MempoolPolicy::Shareable) == Result::Success);
+    REQUIRE(canvas->mempool(SwCanvas::MempoolPolicy::Default) == Result::Success);
+    REQUIRE(canvas->mempool(SwCanvas::MempoolPolicy::Individual) == Result::Success);
+    REQUIRE(canvas->mempool(SwCanvas::MempoolPolicy::Shareable) == Result::Success);
+
```

Comment:
I understand, that it's not the place to check, what will happen if we try to change the mempool policy after pushing something into canvas, right? but if not now then when?

Reply:
@mgrudzinska it will return InsufficientConditin as the doc say.

    /**
     * @brief Set sw engine memory pool behavior policy.
     *
     * Basically ThorVG draws a lot of shapes, it allocates/deallocates a few chunk of memory
     * while processing rendering. It internally uses one shared memory pool
     * which can be reused among the canvases in order to avoid memory overhead.
     *
     * Thus ThorVG suggests memory pool policy to satisfy user demands,
     * if it needs to guarantee the thread-safety of the internal data access.
     *
     * @param[in] policy Use the shared cache memory. The default value is @c true
     *
     * @retval Result::Success When succeed.
     * @retval Result::InsufficientCondition If the canvas has any paints.
     * @retval Result::NonSupport In case the software engine is not supported.
     *
     * @note When @c policy is set as @c MempoolPolicy::Individual, current instance of canvas uses its own individual
     *       memory data that is not shared with others. This is necessary when the canvas is accessed on a worker-thread.
     *
     * @warning It's not allowed after pushing any paints.
     *
     * @BETA_API
    */

Reply:
yes, I know what will happen, I just asked whether we want to check that in the test, but I understand that we want.

Reply:
@mgrudzinska good catch. all scenarios as we can imagine should be covered. unfortunately this patch is broken by my mistake but i will update with new one.